app/src/main/java/com/github/geohunt/app/sensor/PermissionUtils.kt

```diff
@@ -1,194 +1,98 @@
- * then manually call the launchPermissionRequest on the resulting object
- */
+@OptIn(ExperimentalPermissionsApi::class)
 @Composable
-fun rememberPermissionsState(vararg permissions: String): MultiplePermissionState {
-    val mutableMultiplePermissionState = remember {
-        MutableMultiplePermissionState()
-    }
-    mutableMultiplePermissionState.Initialize(permissions.asList())
-    return mutableMultiplePermissionState
+fun RequireFineLocationPermissions(withPermission: @Composable () -> Unit) {
+    RequirePermissions(
+        permissions = listOf(
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.ACCESS_COARSE_LOCATION
+        ),
+        withoutPermission = { locationPermissionsState ->
+            val textToShow = stringResource(R.string.require_location_permission)
+            val buttonText = stringResource(R.string.require_location_permission_button)
+
+            ShowPermissionRequestPage(locationPermissionsState, textToShow, buttonText)
+
+        }, withPermission
+    )
 }
 
-
-private class MutableMultiplePermissionState() : MultiplePermissionState {
-
-    override lateinit var permissions: MutableState<List<Permission>>
-    var launcher: ActivityResultLauncher<Array<String>>? = null
-    var future: CompletableFuture<Void>? = null
-
-    @Composable
-    fun Initialize(perms : List<String>) {
-        val context = LocalContext.current
-
-        permissions = remember {
-            mutableStateOf(perms.map {
-                Permission(it).updated(context)
-            })
-        }
-
-        launcher = rememberLauncherForActivityResult(
-            ActivityResultContracts.RequestMultiplePermissions()
-        ) { result ->
-            updatePermissions(result)
-        }
-    }
-
-    private fun updatePermissions(permissionsUpdated: Map<String, Boolean>) {
-        // First update list of the mutable permission state
-        permissions.value =
-            permissions.value.map {
-                if (permissionsUpdated.getOrElse(it.name) { true }) {
-                    it.updated(Granted)
-                } else {
-                    it.updated(Denied(true))
-                }
+@OptIn(ExperimentalPermissionsApi::class)
+@Composable
+private fun ShowPermissionRequestPage(
+    locationPermissionsState: MultiplePermissionsState,
+    textToShow: String,
+    buttonText: String
+) {
+    Box(
+        contentAlignment = Alignment.Center,
+        modifier = Modifier.fillMaxSize()
+    ) {
+        Column(
+            horizontalAlignment = Alignment.CenterHorizontally,
+            verticalArrangement = Arrangement.Center,
+            modifier = Modifier.align(Alignment.Center)
+                .padding(20.dp, 0.dp),
+        ) {
+            LaunchedEffect(Unit) {
+                Log.i("GeoHunt", "Launching permission request")
+                locationPermissionsState.launchMultiplePermissionRequest()
             }

```

Comment:
Avoid using `LaunchedEffect` just for the sake of launching coroutines. It should be done in the `ViewModel` instead.

Reply:
launchMultiplePermissionRequest is not a coroutine